flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java

```diff
@@ -103,12 +103,28 @@ public void open() throws Exception {
 		this.format.setRuntimeContext(getRuntimeContext());
 		this.format.configure(new Configuration());
 
-		this.collector = new TimestampedCollector<>(output);
 		this.checkpointLock = getContainingTask().getCheckpointLock();
 
 		Preconditions.checkState(reader == null, "The reader is already initialized.");
 
-		this.reader = new SplitReader<>(format, serializer, collector, checkpointLock, readerState);
+		// set the reader context based on the time characteristic
```

Comment:
I think both the `SourceContext` plus subclasses and this instantiation code should be moved out of the sources since it is now used for more than that.